File: client/src/sovd_client.cpp
#include "sovd/client/sovd_client.hpp"

#include <chrono>
#include <thread>

namespace sovd::client {

using json = nlohmann::json;

namespace {

DataValue parse_data_value(const json &d) {
    DataValue v;
    v.id = d.value("id", "");
    if (d.contains("did")) v.did = d["did"].get<std::string>();
    if (d.contains("value")) v.value = d["value"];
    if (d.contains("unit")) v.unit = d["unit"].get<std::string>();
    if (d.contains("error")) v.error = d["error"].get<std::string>();
    if (d.contains("message")) v.error_message = d["message"].get<std::string>();
    return v;
}

std::string join_ids(const std::vector<std::string> &ids) {
    std::string out;
    for (size_t i = 0; i < ids.size(); ++i) {
        if (i) out += ",";
        out += ids[i];
    }
    return out;
}

} // namespace

SovdClient::SovdClient(std::string base_url, RetryPolicy retry) : cli_(base_url), retry_(retry) {
    cli_.set_connection_timeout(3, 0);
    cli_.set_read_timeout(10, 0);
}

json SovdClient::request(const std::string &method, const std::string &url_path, const std::string &body,
                          const std::string &lock_id) {
    httplib::Headers headers;
    if (!lock_id.empty()) headers.emplace("X-SOVD-Lock-Id", lock_id);

    int backoff_ms = retry_.backoff_ms;
    for (int attempt = 0;; ++attempt) {
        httplib::Result res;
        if (method == "GET") {
            res = cli_.Get(url_path, headers);
        } else if (method == "POST") {
            res = cli_.Post(url_path, headers, body, "application/json");
        } else if (method == "PUT") {
            res = cli_.Put(url_path, headers, body, "application/json");
        } else if (method == "DELETE") {
            res = cli_.Delete(url_path, headers);
        } else {
            throw std::invalid_argument("unsupported method: " + method);
        }

        if (!res) {
            throw SovdError(0, "TRANSPORT", "request failed: " + httplib::to_string(res.error()));
        }

        // Retry only 503/504, never 423 -- retrying a lock conflict just
        // hammers whoever actually holds it (CLAUDE.md).
        if ((res->status == 503 || res->status == 504) && attempt < retry_.max_retries) {
            std::this_thread::sleep_for(std::chrono::milliseconds(backoff_ms));
            backoff_ms *= 2;
            continue;
        }

        if (res->status < 200 || res->status >= 300) {
            std::string code = "HTTP_" + std::to_string(res->status);
            std::string message = res->body;
            try {
                json err = json::parse(res->body);
                if (err.contains("error")) code = err["error"].get<std::string>();
                if (err.contains("message")) message = err["message"].get<std::string>();
            } catch (...) {
                // body wasn't the usual {"error","message"} shape -- keep the raw body as the message
            }
            throw SovdError(res->status, code, message);
        }

        if (res->body.empty()) return json::object();
        try {
            return json::parse(res->body);
        } catch (...) {
            throw SovdError(res->status, "BAD_RESPONSE", "server returned a non-JSON body");
        }
    }
}

std::vector<EntityInfo> SovdClient::list_entities() {
    json j = request("GET", "/v1/entities");
    std::vector<EntityInfo> out;
    for (auto &item : j.at("items")) {
        out.push_back({item.value("path", ""), item.value("type", ""), item.value("has_backend", false)});
    }
    return out;
}

DocsResult SovdClient::get_docs(const std::string &path) {
    json j = request("GET", "/v1/entities/" + path + "/docs");

    DocsResult out;
    out.path = j.value("path", path);
    out.type = j.value("type", "");
    out.has_backend = j.value("has_backend", false);

    if (j.contains("capabilities")) {
        const json &c = j["capabilities"];
        Capabilities cap;
        cap.supports_batch_read = c.value("supports_batch_read", false);
        cap.supports_async_operations = c.value("supports_async_operations", false);
        cap.supports_io_control = c.value("supports_io_control", false);
        out.capabilities = cap;
    }

    for (auto &d : j.value("data", json::array())) {
        DataItemDoc item;
        item.id = d.value("id", "");
        item.did = d.value("did", "");
        item.type = d.value("type", "");
        item.access = d.value("access", "");
        item.io_control = d.value("io_control", false);
        if (d.contains("requires_session")) item.requires_session = d["requires_session"].get<std::string>();
        if (d.contains("length")) item.length = d["length"].get<int>();
        if (d.contains("scale")) item.scale = d["scale"].get<double>();
        if (d.contains("unit")) item.unit = d["unit"].get<std::string>();
        if (d.contains("values")) item.values = d["values"];
        out.data.push_back(std::move(item));
    }

    for (auto &o : j.value("operations", json::array())) {
        OperationDoc op;
        op.id = o.value("id", "");
        op.routine_id = o.value("routine_id", "");
        op.async = o.value("async", false);
        if (o.contains("requires_session")) op.requires_session = o["requires_session"].get<std::string>();
        out.operations.push_back(std::move(op));
    }

    return out;
}

std::vector<FaultInfo> SovdClient::get_faults(const std::string &path, const std::string &status_filter) {
    std::string url = "/v1/entities/" + path + "/faults";
    if (!status_filter.empty()) url += "?status=" + status_filter;
    json j = request("GET", url);
    std::vector<FaultInfo> out;
    for (auto &f : j.at("faults")) out.push_back({f.value("code", ""), f.value("status", "")});
    return out;
}

void SovdClient::clear_faults(const std::string &path, const std::string &lock_id) {
    request("DELETE", "/v1/entities/" + path + "/faults", "", lock_id);
}

DataValue SovdClient::get_data(const std::string &path, const std::string &id) {
    return parse_data_value(request("GET", "/v1/entities/" + path + "/data/" + id));
}

std::vector<DataValue> SovdClient::get_data_batch(const std::string &path, const std::vector<std::string> &ids) {
    json j = request("GET", "/v1/entities/" + path + "/data?ids=" + join_ids(ids));
    std::vector<DataValue> out;
    for (auto &item : j.at("items")) out.push_back(parse_data_value(item));
    return out;
}

void SovdClient::put_data(const std::string &path, const std::string &id, const std::string &hex_value,
                           const std::string &lock_id) {
    request("PUT", "/v1/entities/" + path + "/data/" + id, json{{"value", hex_value}}.dump(), lock_id);
}

void SovdClient::set_mode(const std::string &path, const std::string &mode, const std::string &lock_id) {
    request("POST", "/v1/entities/" + path + "/modes", json{{"mode", mode}}.dump(), lock_id);
}

json SovdClient::execute_operation(const std::string &path, const std::string &op, const std::string &params_json,
                                    const std::string &lock_id) {
    return request("POST", "/v1/entities/" + path + "/operations/" + op, params_json.empty() ? "{}" : params_json,
                    lock_id);
}

std::string SovdClient::acquire_lock(const std::string &path, int ttl_seconds) {
    json j = request("POST", "/v1/entities/" + path + "/locks", json{{"ttl_seconds", ttl_seconds}}.dump());
    return j.at("lock_id").get<std::string>();
}

void SovdClient::renew_lock(const std::string &path, const std::string &lock_id, int ttl_seconds) {
    request("PUT", "/v1/entities/" + path + "/locks/" + lock_id, json{{"ttl_seconds", ttl_seconds}}.dump());
}

void SovdClient::release_lock(const std::string &path, const std::string &lock_id) {
    request("DELETE", "/v1/entities/" + path + "/locks/" + lock_id);
}

} // namespace sovd::client

File: web/src/api/sovdClient.ts
// Phase 7: browser SDK, mirroring client/include/sovd/client/sovd_client.hpp's
// shape (same wire format, same rules) rather than reinventing them:
// - Discover the API version from `/` instead of hardcoding `/v1/` (CLAUDE.md:
//   "/" is deliberately unversioned so a client can read api_versions before
//   it knows which prefix to use).
// - Retry only 503/504, never 423 (CLAUDE.md: retrying a lock conflict just
//   hammers whoever holds it).

export interface EntityInfo {
  path: string
  type: string
  has_backend: boolean
}

export interface FaultInfo {
  code: string
  status: string
}

export interface DataItemDoc {
  id: string
  did: string
  type: 'string' | 'float' | 'enum' | 'raw'
  access: 'read' | 'write' | 'read_write'
  io_control?: boolean
  requires_session?: string
  length?: number
  scale?: number
  unit?: string
  values?: Record<string, string>
}

export interface OperationDoc {
  id: string
  routine_id: string
  async: boolean
  requires_session?: string
}

export interface Capabilities {
  supports_batch_read: boolean
  supports_async_operations: boolean
  supports_io_control: boolean
}

export interface DocsResult {
  path: string
  type: string
  has_backend: boolean
  capabilities?: Capabilities
  data: DataItemDoc[]
  operations: OperationDoc[]
}

export interface DataValue {
  id: string
  did?: string
  value?: unknown
  unit?: string
  error?: string
  message?: string
}

export class SovdError extends Error {
  status: number
  code: string
  constructor(status: number, code: string, message: string) {
    super(`${code}: ${message}`)
    this.status = status
    this.code = code
  }
}

export class SovdClient {
  apiBase: string | null = null
  readonly baseUrl: string
  // SOVD_REVIEW_FEEDBACK.md Task 1b: optional OAuth2 bearer token. Empty
  // (the default) sends no Authorization header at all, so the no-auth demo
  // path this project started with keeps working unchanged -- same
  // "unset means no check" shape as the server's own SOVD_OAUTH2_SECRET.
  token: string | null = null

  constructor(baseUrl: string) {
    this.baseUrl = baseUrl
  }

  async ensureApiBase(): Promise<string> {
    if (this.apiBase) return this.apiBase
    const res = await fetch(`${this.baseUrl}/`)
    if (!res.ok) throw new SovdError(res.status, 'ROOT_UNREACHABLE', 'failed to reach server root')
    const body = await res.json()
    const version: string = (body.api_versions ?? [])[0] ?? 'v1'
    this.apiBase = `${this.baseUrl}/${version}`
    return this.apiBase
  }

  async request(
    method: string,
    path: string,
    opts: { body?: unknown; lockId?: string } = {},
  ): Promise<any> {
    const base = await this.ensureApiBase()
    const headers: Record<string, string> = {}
    if (opts.body !== undefined) headers['Content-Type'] = 'application/json'
    if (opts.lockId) headers['X-SOVD-Lock-Id'] = opts.lockId
    if (this.token) headers['Authorization'] = `Bearer ${this.token}`

    let backoffMs = 200
    for (let attempt = 0; ; attempt++) {
      const res = await fetch(base + path, {
        method,
        headers,
        body: opts.body !== undefined ? JSON.stringify(opts.body) : undefined,
      })
      if ((res.status === 503 || res.status === 504) && attempt < 3) {
        await new Promise((r) => setTimeout(r, backoffMs))
        backoffMs *= 2
        continue
      }
      if (!res.ok) {
        const raw = await res.text()
        let code = `HTTP_${res.status}`
        let message = raw
        try {
          const err = JSON.parse(raw)
          code = err.error ?? code
          message = err.message ?? message
        } catch {
          // body wasn't the usual {error,message} shape -- keep the raw text
        }
        throw new SovdError(res.status, code, message)
      }
      const text = await res.text()
      return text ? JSON.parse(text) : {}
    }
  }

  async listEntities(): Promise<EntityInfo[]> {
    return (await this.request('GET', '/entities')).items
  }

  async getDocs(path: string): Promise<DocsResult> {
    return this.request('GET', `/entities/${path}/docs`)
  }

  async getFaults(path: string, statusFilter = ''): Promise<FaultInfo[]> {
    const qs = statusFilter ? `?status=${encodeURIComponent(statusFilter)}` : ''
    return (await this.request('GET', `/entities/${path}/faults${qs}`)).faults
  }

  async clearFaults(path: string, lockId: string): Promise<void> {
    await this.request('DELETE', `/entities/${path}/faults`, { lockId })
  }

  async getDataBatch(path: string, ids: string[]): Promise<DataValue[]> {
    const qs = ids.map(encodeURIComponent).join(',')
    return (await this.request('GET', `/entities/${path}/data?ids=${qs}`)).items
  }

  async putData(path: string, id: string, value: unknown, lockId: string): Promise<void> {
    await this.request('PUT', `/entities/${path}/data/${id}`, { body: { value }, lockId })
  }

  async acquireLock(path: string, ttlSeconds: number): Promise<string> {
    return (await this.request('POST', `/entities/${path}/locks`, { body: { ttl_seconds: ttlSeconds } })).lock_id
  }

  async renewLock(path: string, lockId: string, ttlSeconds: number): Promise<void> {
    await this.request('PUT', `/entities/${path}/locks/${lockId}`, { body: { ttl_seconds: ttlSeconds } })
  }

  async releaseLock(path: string, lockId: string): Promise<void> {
    await this.request('DELETE', `/entities/${path}/locks/${lockId}`)
  }

  // Best-effort, synchronous-dispatch release for a `beforeunload` handler --
  // no retry, doesn't await the response, uses fetch's `keepalive` flag so
  // the request can survive the page tearing down (the modern replacement
  // for sendBeacon, which is POST-only and can't do DELETE). Silently does
  // nothing if the API base was never resolved (no lock could exist yet).
  releaseLockBestEffort(path: string, lockId: string): void {
    if (!this.apiBase) return
    const headers: Record<string, string> = { 'X-SOVD-Lock-Id': lockId }
    if (this.token) headers['Authorization'] = `Bearer ${this.token}`
    fetch(`${this.apiBase}/entities/${path}/locks/${lockId}`, {
      method: 'DELETE',
      headers,
      keepalive: true,
    }).catch(() => {})
  }

  // SOVD_REVIEW_FEEDBACK.md Task 1c: EventSource cannot set an Authorization
  // header, so the stream endpoint can't be gated the same way every other
  // route is once OAuth2 is on. When a token is configured, mint a
  // short-lived single-use ticket through a normal bearer-authenticated
  // POST first and carry *that* in the query string instead of the token
  // itself -- 30s TTL, burned on first use, so it's useless the moment the
  // stream has actually opened. No-op (no ticket param at all) when no
  // token is set, matching every other "unset means no check" seam in this
  // project -- the plain no-auth demo path is unaffected.
  async streamUrl(path: string, id: string, intervalMs: number): Promise<string> {
    const base = await this.ensureApiBase()
    let ticketParam = ''
    if (this.token) {
      const ticket = await this.request('POST', `/entities/${path}/data/${id}/stream-ticket`)
      ticketParam = `&ticket=${encodeURIComponent(ticket.ticket)}`
    }
    return `${base}/entities/${path}/data/${id}/stream?interval_ms=${intervalMs}${ticketParam}`
  }
}
